homeassistant/components/smartthings/smartapp.py

```diff
@@ -0,0 +1,266 @@
+    from pysmartthings import App, AppOAuth, AppSettings
+    from pysmartapp.const import SETTINGS_APP_ID
+
+    # Create app from template attributes
+    template = _get_app_template(hass)
+    app = App()
+    for key, value in template.items():
+        setattr(app, key, value)
+    app = (await api.create_app(app))[0]
+    _LOGGER.debug("Created SmartApp '%s' (%s).", app.app_name, app.app_id)
+
+    # Set unique hass id in settings
+    settings = AppSettings(app.app_id)
+    settings.settings[SETTINGS_APP_ID] = app.app_id
+    settings.settings[SETTINGS_INSTANCE_ID] = \
+        hass.data[DOMAIN][CONF_INSTANCE_ID]
+    await api.update_app_settings(settings)
+    _LOGGER.debug("Updated App Settings for SmartApp '%s' (%s).",
+                  app.app_name, app.app_id)
+
+    # Set oauth scopes
+    oauth = AppOAuth(app.app_id)
+    oauth.client_name = 'Home Assistant'
+    oauth.scope.extend(APP_OAUTH_SCOPES)
+    await api.update_app_oauth(oauth)
+    _LOGGER.debug("Updated App OAuth for SmartApp '%s' (%s).",
+                  app.app_name, app.app_id)
+    return app
+
+
+async def update_app(hass: HomeAssistantType, app):
+    """Ensure the SmartApp is up-to-date and update if necessary."""
+    template = _get_app_template(hass)
+    template.pop('app_name')  # don't update this
+    update_required = False
+    for key, value in template.items():
+        if getattr(app, key) != value:
+            update_required = True
+            setattr(app, key, value)
+    if update_required:
+        await app.save()
+        _LOGGER.debug("SmartApp '%s' (%s) updated with latest settings.",
+                      app.app_name, app.app_id)
+
+
+def setup_smartapp(hass, app):
+    """
+    Configure an individual SmartApp in hass.
+
+    Register the SmartApp with the SmartAppManager so that hass will service
+    lifecycle events (install, event, etc...).  A unique SmartApp is created
+    for each SmartThings account that is configured in hass.
+    """
+    manager = hass.data[DOMAIN][DATA_MANAGER]
+    smartapp = manager.smartapps.get(app.app_id)
+    if smartapp:
+        # already setup
+        return smartapp
+    smartapp = manager.register(app.app_id, app.webhook_public_key)
+    smartapp.name = app.display_name
+    smartapp.description = app.description
+    smartapp.permissions.extend(APP_OAUTH_SCOPES)
+    return smartapp
+
+
+async def setup_smartapp_endpoint(hass: HomeAssistantType):
+    """
+    Configure the SmartApp webhook in hass.
+
+    SmartApps are an extension point within the SmartThings ecosystem and
+    is used to receive push updates (i.e. device updates) from the cloud.
+    """
+    from pysmartapp import Dispatcher, SmartAppManager
+
+    data = hass.data.get(DOMAIN)
+    if data:
+        # already setup
+        return
+
+    # Get/create config to store a unique id for this hass instance.
+    path = hass.config.path(SMARTTHINGS_CONFIG_FILE)
+    if not await hass.async_add_job(os.path.isfile, path):
```

Comment:
Why not store this info inside the config entry?

And if you really need to use a file, use `helpers.storage.Store`

Reply:
It's global to the component (regardless of the number of instances setup) and needed during config flow setup (prior to any entry).  I'll explore using `Store`.

Reply:
We can generate/register it during the config flow and store it inside a config entry when it is created.

If it has to be shared among config entries, then do we really need multiple config entries?

Reply:
It's at a higher level than a config entry.  Yes they are needed.  Consider:
- This component supports _N_ SmartThings accounts.
    - A webhook URL is defined at the component level
- An account has _N_ SmartApps:
    - We create one SmartApp per account setup providing the webhook URL.  Effectively, the data model is:
 `HASS n-n Access Token n-1 Account 1-1 SmartApp`
- Accounts also have _N_ locations.
- Users 'Install' an _instance_ of a SmartApps (called `installed app`) into _one or more_ locations via the SmartThings mobile app.  This is what starts sending data to the end-point, but only within the scope of that specific location.
- **Config entries represent the `installed app` (SmartApp instance).**  Access Token (account) by location and SmartApp.

SmartThings took a simple concept like 'webhooks' and wrapped it into an extremely complex construct.  I can understand the confusion...  :) The links in the PR description take you to the documentation for how all of this works.